### DISCUSSION TITLE
Save_model and load_model function don't save/load optimizer state

### DISCUSSION THREAD

**Discussion by ghost** [artefactory/choice-learn#278] [General]
I was trying to save and load a ConditionalLogit model for resuming a training, but the implementation inside the base class ChoiceModel doesn't save and load the optimizer state, which it's essential when resuming a training.

**Comment by VincentAuriau** [comment]
Let me check if I can quickly come up with a solution for this . I will update you soon.

For the time being I suppose you can re instantiate the optimizer - you may lose a few epochs but I suppose the results should be ok.

**Reply by ghost** [reply]
And also, saving model checkpoints (e.g. every _x_ epochs) in the fit function would be very useful during training. Especially when dealing with very large datasets, if we encounter any issue during training like node failures, it would help to be able to resume from a checkpoint.

**Reply by VincentAuriau** [reply]
I have a first version that looks "ok" [here](https://github.com/artefactory/choice-learn/tree/save-opt). I am still running a few checks to be sure.

For the checkpoints during the training you can use a callback. I am writing an example in a notebook, will let you know when I push it.

Thanks !

**Reply by VincentAuriau** [reply]
[Here](https://github.com/artefactory/choice-learn/blob/save-opt/notebooks/models/saving_loading_models.ipynb) is simple example with a callback to save a model every n epochs.

If you want fancier options you can check the [official implementation](https://github.com/keras-team/keras/blob/v3.3.3/keras/src/callbacks/model_checkpoint.py#L14-L416) from tensorflow  and copy some stuff

**Reply by ghost** [reply]
Thank you! :) Appreciate it a lot!

**Reply by ghost** [reply]
> I have a first version that looks "ok" [here](https://github.com/artefactory/choice-learn/tree/save-opt). I am still running a few checks to be sure.
> 
> For the checkpoints during the training you can use a callback. I am writing an example in a notebook, will let you know when I push it.
> 
> Thanks !

Just tested this out (save_model) function: 

1. I get a warning - but I see that the trainable_weights are saved above this warning so I assume we can ignore it :
WARNING:root:Attribute '_trainable_weights' is a list with non-serializable
                         types and will not be saved. 
2. If I call the save_model one more time and I already have the 'optimizer' folder created from a previous run, it will fail (line 553 in base_model.py), I think adding "exist_ok=True" will solve the issue.
3. I get an error when trying to save the learning rate (line 560 in base_model.py), because my learning rate it's float and cannot be converted to list

**Reply by VincentAuriau** [reply]
1. Yes you can ignore the message (and I removed it)
2. just commited this - I suppose I will later add a warning message that the saved model is overwritten
3. Is it a numpy error that is raised ? Can you copy the exact error please?

And thanks a lot for your feedbacks !

**Reply by ghost** [reply]
3. Sure. Here it is:
<img width="619" height="132" alt="image" src="https://github.com/user-attachments/assets/cfbe98a7-04aa-4318-997f-df293cef3159" />

config['learning_rate'] it's simply a float value in my case:
<img width="432" height="439" alt="image" src="https://github.com/user-attachments/assets/780eaee1-6f8e-43ee-a0ae-4be17477c3e3" />

I simply commented the lines where it converts .tolist() and now it works.

**Reply by ghost** [reply]
> [Here](https://github.com/artefactory/choice-learn/blob/save-opt/notebooks/models/saving_loading_models.ipynb) is simple example with a callback to save a model every n epochs.
> 
> If you want fancier options you can check the [official implementation](https://github.com/keras-team/keras/blob/v3.3.3/keras/src/callbacks/model_checkpoint.py#L14-L416) from tensorflow and copy some stuff

This works perfectly ;) I managed to save the checkpoints. But one question is when I want to resume and use e.g. an model from epoch 8, shouldn't the epoch number in the fit() function also be resumed starting from the one that was saved (line 289 in base_model.py - the tqdm,trange starts at position 0 rather than resuming from the epoch it was left). I'm attaching here an example: 
<img width="685" height="729" alt="image" src="https://github.com/user-attachments/assets/10a7d633-1d54-4222-9d41-081272884b69" />
I would've expected that the epochs start from epoch 9 and continue until epoch 15

**Reply by VincentAuriau** [reply]
I pushed a fix - it looks like depending on the tf version, the configs returned have different types - let me know if it still does not work

**Reply by VincentAuriau** [reply]
> > [Here](https://github.com/artefactory/choice-learn/blob/save-opt/notebooks/models/saving_loading_models.ipynb) is simple example with a callback to save a model every n epochs.
> > If you want fancier options you can check the [official implementation](https://github.com/keras-team/keras/blob/v3.3.3/keras/src/callbacks/model_checkpoint.py#L14-L416) from tensorflow and copy some stuff
> 
> This works perfectly ;) I managed to save the checkpoints. But one question is when I want to resume and use e.g. an model from epoch 8, shouldn't the epoch number in the fit() function also be resumed starting from the one that was saved (line 289 in base_model.py - the tqdm,trange starts at position 0 rather than resuming from the epoch it was left). I'm attaching here an example: <img alt="image" width="685" height="729" src="https://private-user-images.githubusercontent.com/236864042/500548455-10a7d633-1d54-4222-9d41-081272884b69.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjAzNzExMTUsIm5iZiI6MTc2MDM3MDgxNSwicGF0aCI6Ii8yMzY4NjQwNDIvNTAwNTQ4NDU1LTEwYTdkNjMzLTFkNTQtNDIyMi05ZDQxLTA4MTI3Mjg4NGI2OS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMDEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTAxM1QxNTUzMzVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yNDA0NmNhYzY4YjBmZTZkMzNmOGNmYmM0Mjc5MjA3NDc5NzI5NzgyMDE2NDA2YjZhOWU4ODNiYjMyZmQzMGM0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.9KV56HAqvZ8JKSRujggg58s2JtS7mlK5OSMTPFcH5ME"> I would've expected that the epochs start from epoch 9 and continue until epoch 15

It is a question I had. Usually the number of epochs is given directly in the fit. For simplicity I added it in the ```__init__ ``` so that it can be saved as an hyper parameter. I kinda kept this behavior so that the model wan be trained several times if needs be (without having to change the n_epochs argument).

I can add an argument 'total_trained_epochs' or something similar that might help you know how many additional epochs you need.
